[WIP] Improve metrics UI with charts

adamlogic opened the pull request:
This is a work-in-progress to address #5458. It introduces [Chart.js](https://www.chartjs.org/docs/latest/) to the new metrics page, plotting the total execution time of the 10 most time-consuming jobs on a line chart.

<img width="1134" alt="CleanShot 2022-08-10 at 16 37 06@2x" src="https://user-images.githubusercontent.com/372/184015341-97e441fc-861a-4e2f-a6d7-6c340b20b30a.png">

adamlogic commented on lib/sidekiq/metrics/query.rb:
```diff
@@ -35,48 +35,32 @@ def initialize(pool: Sidekiq.redis_pool, now: Time.now)
-        time = @time
+      # Get metric data for all jobs from the last hour
+      def top_jobs(minutes: 60)
+        result = Result.new
```
I had a lot of trouble wrapping my head around the previous implementation and shaping the data the way I needed it for the chart. It helped me to extract the `Query::Result` and `Query::JobResult` classes here, but I'm sure we could also get it working with the nested Hashes if you don't like this approach.

adamlogic commented on lib/sidekiq/metrics/query.rb:
```diff
@@ -35,48 +35,32 @@ def initialize(pool: Sidekiq.redis_pool, now: Time.now)
+            minutes.times do |idx|
               key = "j|#{time.strftime("%Y%m%d")}|#{time.hour}:#{time.min}"
               pipe.hgetall key
+              result.prepend_bucket time
```
We need this running tally of all the "buckets" so we can provide the chart a complete x-axis, even for times when there's no job data.

adamlogic commented on test/test_metrics.rb:
```diff
@@ -104,16 +99,19 @@ def create_known_metrics(time = fixed_time)
+      assert_equal %w[App::SomeJob App::FooJob].sort, result.job_results.keys.sort
+      some_job_result = result.job_results["App::SomeJob"]
+      refute_nil some_job_result
+      assert_equal 1, some_job_result.series.dig("ms", "22:03")
```
This test is randomly failing with different actual values, and I'm not sure why, since we're using a fixed time. It passes most of the time.

```
Failure:
Sidekiq::Metrics::querying#test_0002_fetches top job data [test/test_metrics.rb:118]:
Expected: 1
  Actual: 8

Failure:
Sidekiq::Metrics::querying#test_0002_fetches top job data [test/test_metrics.rb:118]:
Expected: 1
  Actual: 2
```

davidwessman replied:
Could it help to debug it by adding the whole hash as the message?

```suggestion
      assert_equal 1, some_job_result.series.dig("ms", "22:03"), some_job_result
```

adamlogic replied:
Fixed in https://github.com/mperham/sidekiq/pull/5467/commits/d0b1616ad82250204cf1a270aac17a64000de37b.

adamlogic commented on web/assets/javascripts/metrics.js:
```diff
@@ -0,0 +1,41 @@
+      },
+    },
+    interaction: {
+      mode: "x",
```
This configures the tooltip to show all points at the current spot on the x-axis, rather than just a single datapoint.

<img width="184" alt="CleanShot 2022-08-10 at 16 47 53@2x" src="https://user-images.githubusercontent.com/372/184017209-188cc192-d656-4901-afac-b004315750af.png">

versus...

<img width="206" alt="CleanShot 2022-08-10 at 16 48 25@2x" src="https://user-images.githubusercontent.com/372/184017233-d1de183e-760d-4f83-96d4-a965c0eb37ea.png">

adamlogic commented on web/assets/javascripts/metrics.js:
```diff
@@ -0,0 +1,41 @@
+    },
+    plugins: {
+      legend: {
+        display: false,
```
I opted the hide the default legend because I thought it was ugly. I think the swatches in the data table are just as effective and look a lot nicer.

![CleanShot 2022-08-10 at 16 49 21@2x](https://user-images.githubusercontent.com/372/184017482-99ce6444-fcb6-420f-8a94-3d471895ce69.png)

We could also potentially make the data table swatches interactive if, for example, we only wanted to initially show the top 5 job classes in the chart. You could use the swatch to toggle a job class' visibility in the chart. 🤷‍♂️

adamlogic commented on web/assets/javascripts/metrics.js:
```diff
@@ -0,0 +1,41 @@
+  });
+}
+
+document.addEventListener("DOMContentLoaded", init);
```
The dashboard uses this [custom `ready` function](https://github.com/adamlogic/sidekiq/blob/3ab19ed514c10510e4e403067fcd61e5f2b48164/web/assets/javascripts/dashboard.js#L224-L227), but I don't know if it's necessary.

mperham replied:
I don't know either. If it's not needed, it's a bug.

adamlogic commented on web/views/metrics.erb:
```diff
@@ -1,50 +1,58 @@
+<script type="text/javascript" src="<%= root_path %>javascripts/metrics.js"></script>
 
-<h1><%= t('Metrics') %></h1>
+<h3><%= t('Metrics') %></h3>
```
Changed this heading level to match other pages in the web UI.

adamlogic commented on web/views/metrics.erb:
```diff
@@ -1,50 +1,58 @@
+></canvas>
 
-<h3>Top Jobs by Execution Time</h3>
+<h3>Most Time-Consuming Jobs</h3>
```
Renamed this in case we want to make this table sortable. It will also list the top X most time-consuming jobs, but they might not be _sorted by_ execution time (although they are right now).

adamlogic reviewed (commented):
Consider this a first draft. Happy to iterate on it as much as we need.

Some things I'm still thinking about...

- [x] Add deploy markers to the chart.
- [ ] Add the ability to change the metric shown in the chart (processed jobs, failed jobs, avg exec time)
- [ ] Show *more* job classes in the data table and *fewer* job classes in the chart. Right now the chart feels a bit too busy and the table a bit too anemic.

mperham commented:
🤩 Holy cow.

Maybe we could have checkboxes in the table rows which show/hide each job type within the graph? And have the graph only show the top 5 by default?

davidwessman commented:
Is it possible to add labels for the x- and y-axis? `Total execution time (s)` vs `Time (hh:mm)`

adamlogic commented:
> Maybe we could have checkboxes in the table rows which show/hide each job type within the graph? And have the graph only show the top 5 by default?

https://github.com/mperham/sidekiq/pull/5467/commits/da81f58146fe4d58abe2fa25c47db9c68550fcdc converted the color swatches to checkboxes, and https://github.com/mperham/sidekiq/pull/5467/commits/da81f58146fe4d58abe2fa25c47db9c68550fcdc wired them up to toggle each job type within the graph.

The top 20 most time-consuming job types will be shown in the data table, but only 5 will initially be shown on the chart.

https://user-images.githubusercontent.com/372/184196628-da4e08a9-1562-45aa-b2e8-de2a029ab880.mp4

We don't have enough data here to trigger the `fallbackColor`, so for the demo below, I deleted all but three colors.


https://user-images.githubusercontent.com/372/184196765-53170c41-a020-4db6-b5b0-515128a1269c.mp4

adamlogic commented:
After posting those demos above, I realized the unchecked checkboxes looked disabled. That's fixed now.

<img width="1200" alt="CleanShot 2022-08-11 at 13 35 00@2x" src="https://user-images.githubusercontent.com/372/184197255-db6b6902-61d6-4f11-8c94-7ab86d12cf74.png">

mperham commented on web/views/metrics.erb:
```diff
@@ -1,50 +1,60 @@
+<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/Chart.js/3.9.1/chart.min.js"></script>
```
We will need to vendor this. The Web UI should be totally self-contained.

mperham replied:
Make sure our copy of chart.js.min includes the MIT license in their repo.

mperham replied:
Oh, looks like it already does include a small license header.

adamlogic replied:
Scripts are now vendored.

mperham reviewed (commented):
Fantastic progress. Any luck on the vertical deploy marks?

adamlogic commented:
UX improvements in https://github.com/mperham/sidekiq/pull/5467/commits/9f5a658f0e83f68d0cededfc09b9380dd80e37ff:

- The chart now displays seconds instead of milliseconds (matches the data table)
- Added a label for the x-axis
- Added rounding and labels to the tooltip

<img width="724" alt="CleanShot 2022-08-11 at 14 04 04@2x" src="https://user-images.githubusercontent.com/372/184208984-a2a30db3-fc1e-4a33-94dc-4ff49bbb80ee.png">

Note that I did not add a label for the x-axis because I didn't think the added noise provided enough value. The x-axis is obviously time, but not obviously in UTC, which is why I added "UTC" to the tooltip.

adamlogic commented:
> Any luck on the vertical deploy marks?

Up next!

adamlogic commented:
Deploy marks!

<img width="1161" alt="CleanShot 2022-08-11 at 16 34 28@2x" src="https://user-images.githubusercontent.com/372/184237452-20dc82d7-c8b4-40e2-88ef-bf6dd024217d.png">

Those labels show the first "word" of the mark label, but the full label is shown in the tooltip.

<img width="1148" alt="CleanShot 2022-08-11 at 16 34 47@2x" src="https://user-images.githubusercontent.com/372/184237529-f66f2ac7-aeee-44a4-95b5-5e2c5bdfe6f6.png">

Perhaps it's too much to assume the first word of the label is the deploy SHA. I'm wondering if we should lose the label and just keep the tooltip.

mperham commented:
That's a reasonable question. We could either add more smarts to the deploy mark by supporting an explicit `scm_label` which could be a git sha, git tag, svn id, etc. Or we could just create a convention where if the first word in the marker label is surrounded by `[word]` then we display `word`. Hard code "smarts" or rely on a convention?

adamlogic commented:
Bug: With the deploy annotations, the chart y-axis scale is no longer updating based on the selected job types. I'll figure it out tomorrow.

https://user-images.githubusercontent.com/372/184244551-7b06c3b6-b505-4e3e-8a72-1a1a3fdb96ce.mp4

adamlogic commented:
> Hard code "smarts" or rely on a convention?

I don't have a strong opinion here, but if it were up to me I'd probably say "neither", and remove the label annotation altogether. It feels like we're trying to assume too much about how this feature is going to be used.

Also, the label annotation required an [ugly hack](https://github.com/mperham/sidekiq/pull/5467/commits/8a8fe3f51aed047e1767c9581c8eab74f365575b) to get the positioning working correctly. The vertical bar didn't require the same hack, so without the label I could remove that hack.

adamlogic commented:
Here's a quick demo of the UX *without* the label.

![CleanShot 2022-08-12 at 08 47 42](https://user-images.githubusercontent.com/372/184357051-f1e6e697-ecda-493c-b8d2-5cc7b69d0d6d.gif)

mperham commented:
Well, an unlabeled line is harder to answer "what is this? why is it here?" Maybe we could have a separate legend or table for deploy marks.

adamlogic commented:
Great! I've removed the label annotations, which really cleaned things up.

How are you feeling about the scope of this PR? I'm not sure there's more I want to do on the metrics overview page, for now at least. I'd rather move on to the job metrics page, and I can either continue with this PR or start a new one.

mperham commented:
I love it as is, I'll test it out and merge if all is well. Open up a new PR for the job specific page.

mperham commented:
If you want extra credit, test it with dark mode and using Hebrew as your browser language (to get RTL UI flow).

mperham commented:
Lots of room for polish but infinitely better than what was there before.
<img width="1206" alt="Screen Shot 2022-08-12 at 9 51 06 AM" src="https://user-images.githubusercontent.com/2911/184406388-a2a97d7e-0a68-40a3-b318-a71142838a0b.png">
